File: packages/expo-updates/android/src/main/java/expo/modules/updates/codesigning/CodeSigningConfiguration.kt
package expo.modules.updates.codesigning

import android.util.Base64
import android.util.Log
import expo.modules.structuredheaders.BooleanItem
import expo.modules.structuredheaders.Dictionary
import expo.modules.structuredheaders.StringItem
import expo.modules.updates.codesigning.CertificateChain.Companion.expoProjectInformation
import java.security.Signature

private const val TAG = "CodeSigning"

enum class ValidationResult {
  VALID,
  INVALID,
  SKIPPED
}

data class SignatureValidationResult(val validationResult: ValidationResult, val expoProjectInformation: ExpoProjectInformation?)

/**
 * Contains all information about code signing.
 * @param embeddedCertificateString Implicitly trusted self-signed root certificate. May be the code signing certificate if only using embedded certificate (no chain).
 * @param codeSigningMetadata Metadata about the code signing certificate. May be the root certificate or a leaf certificate.
 * @param includeManifestResponseCertificateChain Should a certificate chain in the manifest response be evaluated when validating the signature. If true, ignores codeSigningMetadata when evaluating cert chain.
 * @param allowUnsignedManifests Should unsigned manifest responses be allowed. Used during Expo Go transition when not all expo-cli versions know how to sign using the new method yet the new Expo Go clients have certificates.
 */
class CodeSigningConfiguration(
  private val embeddedCertificateString: String,
  private val codeSigningMetadata: Map<String, String>?,
  private val includeManifestResponseCertificateChain: Boolean,
  private val allowUnsignedManifests: Boolean
) {
  private val algorithmFromMetadata: CodeSigningAlgorithm by lazy {
    CodeSigningAlgorithm.parseFromString(
      codeSigningMetadata?.get(
        CODE_SIGNING_METADATA_ALGORITHM_KEY
      )
    )
  }

  private val keyIdFromMetadata: String by lazy {
    codeSigningMetadata?.get(CODE_SIGNING_METADATA_KEY_ID_KEY) ?: CODE_SIGNING_METADATA_DEFAULT_KEY_ID
  }

  fun validateSignature(signature: String?, bodyBytes: ByteArray, manifestResponseCertificateChain: String?): SignatureValidationResult {
    if (signature == null) {
      if (!allowUnsignedManifests) {
        throw Exception("No expo-signature header specified")
      } else {
        // no-op
        return SignatureValidationResult(ValidationResult.SKIPPED, null)
      }
    }

    return validateSignatureInternal(
      SignatureHeaderInfo.parseSignatureHeader(signature),
      bodyBytes,
      manifestResponseCertificateChain
    )
  }

  private fun validateSignatureInternal(info: SignatureHeaderInfo, bodyBytes: ByteArray, manifestResponseCertificateChain: String?): SignatureValidationResult {
    val certificateChain = if (includeManifestResponseCertificateChain) {
      CertificateChain(
        separateCertificateChain(manifestResponseCertificateChain ?: "") + embeddedCertificateString
      )
    } else {
      // check that the key used to sign the response is the same as the key in the code signing certificate
      if (info.keyId != keyIdFromMetadata) {
        throw Exception("Key with keyid=${info.keyId} from signature not found in client configuration")
      }

      // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
      // algorithm specified in the configuration
      if (info.algorithm != algorithmFromMetadata) {
        Log.i(
          TAG,
          "Key with alg=${info.algorithm} from signature does not match client configuration algorithm, continuing"
        )
      }

      CertificateChain(listOf(embeddedCertificateString))
    }

    // For now only SHA256withRSA is supported. This technically should be `metadata.algorithm` but
    // it breaks down when metadata is for a different key than the signing key (the case where intermediate
    // certs are served alongside the manifest and the metadata is for the root embedded cert).
    // In the future if more methods are added we will need to be sure that we think about how to
    // specify what algorithm should be used in the chain case. One approach may be that in the case of
    // chains served alongside the manifest we fork the behavior to trust the `info.algorithm` while keeping
    // `metadata.algorithm` for the embedded case.
    val isValid = Signature.getInstance("SHA256withRSA").apply {
      initVerify(certificateChain.codeSigningCertificate.publicKey)
      update(bodyBytes)
    }.verify(Base64.decode(info.signature, Base64.DEFAULT))

    return SignatureValidationResult(
      if (isValid) ValidationResult.VALID else ValidationResult.INVALID,
      certificateChain.codeSigningCertificate.expoProjectInformation()
    )
  }

  fun getAcceptSignatureHeader(): String {
    return Dictionary.valueOf(
      mapOf(
        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE to BooleanItem.valueOf(true),
        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_KEY_ID to StringItem.valueOf(keyIdFromMetadata),
        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_ALGORITHM to StringItem.valueOf(algorithmFromMetadata.algorithmName)
      )
    ).serialize()
  }

  companion object {
    fun separateCertificateChain(certificateChainInManifestResponse: String): List<String> {
      val startDelimiter = "-----BEGIN CERTIFICATE-----"
      val endDelimiter = "-----END CERTIFICATE-----"
      val certificateStringList = mutableListOf<String>()
      var currStartIndex = 0
      while (true) {
        val startIndex = certificateChainInManifestResponse.indexOf(startDelimiter, currStartIndex)
        val endIndex = certificateChainInManifestResponse.indexOf(endDelimiter, currStartIndex)
        if (startIndex == -1 || endIndex == -1) {
          break
        }
        certificateStringList.add(certificateChainInManifestResponse.substring(startIndex, endIndex + endDelimiter.length))
        currStartIndex = endIndex + endDelimiter.length
      }

      return certificateStringList
    }
  }
}

File: packages/expo-updates/ios/EXUpdates/CodeSigning/CodeSigningConfiguration.swift
// Copyright 2015-present 650 Industries. All rights reserved.

// swiftlint:disable identifier_name

import Foundation
import CommonCrypto

internal struct CodeSigningMetadataFields {
  static let KeyIdFieldKey = "keyid"
  static let AlgorithmFieldKey = "alg"
}

internal enum ValidationResult {
  case valid
  case invalid
  case skipped
}

internal final class SignatureValidationResult {
  private(set) internal var validationResult: ValidationResult
  private(set) internal var expoProjectInformation: ExpoProjectInformation?

  required init(validationResult: ValidationResult, expoProjectInformation: ExpoProjectInformation?) {
    self.validationResult = validationResult
    self.expoProjectInformation = expoProjectInformation
  }
}

@objc(EXUpdatesCodeSigningConfiguration)
public final class CodeSigningConfiguration: NSObject {
  private var embeddedCertificateString: String
  private var keyIdFromMetadata: String
  private var algorithmFromMetadata: CodeSigningAlgorithm
  private var includeManifestResponseCertificateChain: Bool
  private var allowUnsignedManifests: Bool

  internal required init(
    embeddedCertificateString: String,
    metadata: [String: String],
    includeManifestResponseCertificateChain: Bool,
    allowUnsignedManifests: Bool
  ) throws {
    self.embeddedCertificateString = embeddedCertificateString
    self.keyIdFromMetadata = metadata[CodeSigningMetadataFields.KeyIdFieldKey] ?? SignatureHeaderInfo.DefaultKeyId
    self.algorithmFromMetadata = try CodeSigningAlgorithm.parseFromString(metadata[CodeSigningMetadataFields.AlgorithmFieldKey])
    self.includeManifestResponseCertificateChain = includeManifestResponseCertificateChain
    self.allowUnsignedManifests = allowUnsignedManifests
  }

  /**
   * String escaping is defined by https://www.rfc-editor.org/rfc/rfc8941.html#section-3.3.3
   */
  private static func escapeStructuredHeaderStringItem(_ str: String) -> String {
    return str.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: "\"", with: "\\\"")
  }

  internal func createAcceptSignatureHeader() -> String {
    return "sig, keyid=\"\(CodeSigningConfiguration.escapeStructuredHeaderStringItem(keyIdFromMetadata))\", alg=\"\(CodeSigningConfiguration.escapeStructuredHeaderStringItem(algorithmFromMetadata.rawValue))\""
  }

  internal func validateSignature(
    logger: UpdatesLogger,
    signature: String?,
    signedData: Data,
    manifestResponseCertificateChain: String?
  ) throws -> SignatureValidationResult {
    guard let signature = signature else {
      if !self.allowUnsignedManifests {
        throw CodeSigningError.SignatureHeaderMissing
      } else {
        // no-op
        return SignatureValidationResult(validationResult: ValidationResult.skipped, expoProjectInformation: nil)
      }
    }

    return try validateSignatureInternal(
      logger: logger,
      signatureHeaderInfo: try SignatureHeaderInfo.parseSignatureHeader(signatureHeader: signature),
      signedData: signedData,
      manifestResponseCertificateChain: manifestResponseCertificateChain
    )
  }

  private func validateSignatureInternal(
    logger: UpdatesLogger,
    signatureHeaderInfo: SignatureHeaderInfo,
    signedData: Data,
    manifestResponseCertificateChain: String?
  ) throws -> SignatureValidationResult {
    let certificateChain: CertificateChain
    if self.includeManifestResponseCertificateChain {
      certificateChain = try CertificateChain(
        certificateStrings: CodeSigningConfiguration.separateCertificateChain(
          certificateChainInManifestResponse: manifestResponseCertificateChain ?? ""
        ) + [self.embeddedCertificateString]
      )
    } else {
      // check that the key used to sign the response is the same as the key in the code signing certificate
      if signatureHeaderInfo.keyId != self.keyIdFromMetadata {
        throw CodeSigningError.KeyIdMismatchError
      }

      // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
      // algorithm specified in the configuration
      if signatureHeaderInfo.algorithm != self.algorithmFromMetadata {
        logger.warn(message: "Key with alg=\(signatureHeaderInfo.algorithm) from signature does not match client configuration algorithm, continuing")
      }

      certificateChain = try CertificateChain(certificateStrings: [embeddedCertificateString])
    }

    // For now only SHA256withRSA is supported. This technically should be `metadata.algorithm` but
    // it breaks down when metadata is for a different key than the signing key (the case where intermediate
    // certs are served alongside the manifest and the metadata is for the root embedded cert).
    // In the future if more methods are added we will need to be sure that we think about how to
    // specify what algorithm should be used in the chain case. One approach may be that in the case of
    // chains served alongside the manifest we fork the behavior to trust the `info.algorithm` while keeping
    // `metadata.algorithm` for the embedded case.
    let (secCertificate, _) = try certificateChain.codeSigningCertificate()

    guard let publicKey = secCertificate.publicKey else {
      throw CodeSigningError.CertificateMissingPublicKeyError
    }

    guard let signatureData = Data(base64Encoded: signatureHeaderInfo.signature) else {
      throw CodeSigningError.SignatureEncodingError
    }

    let isValid = try self.verifyRSASHA256SignedData(signedData: signedData, signatureData: signatureData, publicKey: publicKey)
    return SignatureValidationResult(
      validationResult: isValid ? ValidationResult.valid : ValidationResult.invalid,
      expoProjectInformation: try certificateChain.codeSigningCertificate().1.expoProjectInformation()
    )
  }

  private func verifyRSASHA256SignedData(signedData: Data, signatureData: Data, publicKey: SecKey) throws -> Bool {
    let hashBytes = signedData.sha256()
    var error: Unmanaged<CFError>?
    if SecKeyVerifySignature(publicKey, .rsaSignatureDigestPKCS1v15SHA256, hashBytes as CFData, signatureData as CFData, &error) {
      return true
    }
    if let error = error, (error.takeRetainedValue() as Error as NSError).code != errSecVerifyFailed {
      throw CodeSigningError.SecurityFrameworkSecKeyVerificationError(cause: error.takeRetainedValue())
    }
    return false
  }

  internal static func separateCertificateChain(certificateChainInManifestResponse: String) -> [String] {
    let startDelimiter = "-----BEGIN CERTIFICATE-----"
    let endDelimiter = "-----END CERTIFICATE-----"
    var certificateStringList = [] as [String]

    var currStartIndex = certificateChainInManifestResponse.startIndex
    while true {
      let startIndex = certificateChainInManifestResponse.firstIndex(of: startDelimiter, startingAt: currStartIndex)
      let endIndex = certificateChainInManifestResponse.firstIndex(of: endDelimiter, startingAt: currStartIndex)

      if let startIndex = startIndex, let endIndex = endIndex {
        let newEndIndex = certificateChainInManifestResponse.index(endIndex, offsetBy: endDelimiter.count)
        certificateStringList.append(String(certificateChainInManifestResponse[startIndex..<newEndIndex]))
        currStartIndex = newEndIndex
      } else {
        break
      }
    }

    return certificateStringList
  }
}

private extension SecCertificate {
  var publicKey: SecKey? {
    SecCertificateCopyKey(self)
  }
}

internal extension OSStatus {
  var isSuccess: Bool {
    self == errSecSuccess
  }
}

private extension Data {
  func sha256() -> Data {
    var digest = Data(count: Int(CC_SHA256_DIGEST_LENGTH))
    withUnsafeBytes { bytes in
      digest.withUnsafeMutableBytes { mutableBytes in
        _ = CC_SHA256(bytes.baseAddress, CC_LONG(count), mutableBytes.bindMemory(to: UInt8.self).baseAddress)
      }
    }
    return digest
  }
}

private extension String {
  func firstIndex(of: String, startingAt: String.Index) -> String.Index? {
    return self[startingAt...].range(of: of)?.lowerBound
  }
}

// swiftlint:enable identifier_name
